README.md
# Lerk

A Discord bot, whose main purpose is to allow quering Hive - Natural Selection
2's ELO system.

Live versions of this bot can be found in the NS2 official Discord server, as
well as servers of independent NS2 communities.

## Running

The public [Docker container](https://hub.docker.com/r/lavode/lerk/) is the preferred way of running this application.

### Configuration

Configuration is done exclusively via environment variables. Wherever possible,
sane defaults have been used.

In the table below, `Required` indicates values you have to define, ie settings
which have no default, but are required for operation.

#### General

| Setting               | Required | Default | Description                                |
| --------------------- | -------- | ------- | ------------------------------------------ |
| `DISCORD_CLIENT_ID`   | y        |         | OAuth Client ID.                           |
| `DISCORD_TOKEN`       | y        |         | OAuth Token.                               |
| `LOG_LEVEL`           | n        | normal  | Logging verbosity. Valid values are 'debug', 'verbose', 'normal', 'quiet', 'silent' |

#### Database

| Setting       | Required | Default   | Description                                                          |
| ------------- | -------- | --------- | -------------------------------------------------------------------- |
| `DB_HOST`     | n        |           | Address of database server. Default is adapter-specific.             |
| `DB_PORT`     | n        |           | Port of database server. Default is adapter-specific.                |
| `DB_DATABASE` | y        |           | Name of database which to use.                                       |
| `DB_USER`     |          |           | User which to authenticate as. Default is adapter-specific.          |
| `DB_PASS`     |          |           | Password with which to authenticate as. Default is adapter-specific. |

#### Lerk

| Setting                  | Required | Default | Description                                |
| ------------------------ | -------- | ------- | ------------------------------------------ |
| `LERK_COMMAND_PREFIX`    | n        | !       | Prefix which is used to identify commands. |
| `LERK_ADMIN_USERS`       | n        |         | Comma-separated list of Discord user IDs which to grant admin access to the bot to. |
| `LERK_HINTS_ADMIN_USERS` | n        |         | Comma-separated list of Discord user IDs which to grant admin access to the ingame hints part of the bot to. |

#### Prometheus

| Setting                | Required | Default | Description                                                 |
| ---------------------- | -------- | ------- | ----------------------------------------------------------- |
| `PROMETHEUS_LISTEN_IP` | n        | 0.0.0.0 | IP which Prometheus exporter will bind to. Might need changing if running outside a Docker container. |
| `PROMETHEUS_PORT`      | n        | 5000    | Port which Prometheus exporter will bind to.                |
| `PROMETHEUS_ENABLED`   | n        | true    | `true` to enable exporter, any other value else to disable. |

#### Hive Interface

| Setting                              | Required | Default | Description                                                 |
| ------------------------------------ | -------- | ------- | ----------------------------------------------------------- |
| `STEAM_WEB_API_KEY`                  | n        |         | API key to use for authentication against Steam web API. Required for resolving of Steam custom URLs. |
| `HIVE_GLOBAL_RATE_LIMIT`             | n        | 2       | Global number of queries to allow in interval.   |
| `HIVE_GLOBAL_RATE_LIMIT_TIME_SPAN`   | n        | 1       | Global rate limiting interval in seconds.        |
| `HIVE_PER_USER_RATE_LIMIT`           | n        | 1       | Per user number of queries to allow in interval. |
| `HIVE_PER_USER_RATE_LIMIT_TIME_SPAN` | n        | 1       | Per user rate limiting interval in seconds.      |
| `HIVE_HELP_MESSAGE_RATE_LIMIT`       | n        | 1       | Per user number of help messages to send in interval. |
| `HELP_MESSAGE_RATE_LIMIT_TIME_SPAN`  | n        | 300     | Per user rate limiting interval in seconds.      |
| `HIVE_ENABLE_EMBEDS`                 | n        | true    | Set to `false` to use plain-text rather than embed-style output. |

#### Observatory

| Setting                | Required | Default                         | Description                       |
| -----------------------| -------- | ------------------------------- | --------------------------------- |
| `OBSERVATORY_BASE_URL` | n        | https://observatory.morrolan.ch | Base URL of observatory instance. |

#### Gorge

| Variable                     | Default value | Required | Description                                               |
| ---------------------------- | ------------- | -------- | --------------------------------------------------------- |
| `GORGE_BASE_URL`             |               | n        | Base URL of Gorge. Empty to disable Gorge integration.    |
| `GORGE_HTTP_BASIC_USER`      |               | n        | User for HTTP basic authentication. Empty to disable.     |
| `GORGE_HTTP_BASIC_PASSWORD ` |               | n        | Password for HTTP basic authentication. Empty to disable. |
| `GORGE_CONNECT_TIMEOUT`      | 1             | y        | HTTP connect timeout towards Gorge API.                   |
| `GORGE_TIMEOUT`              | 2             | y        | HTTP timeout towards Gorge API.                           |
| `STATISTICS_CLASS`           | n_30          | y        | Statistics class of Gorge which to query                  |

#### Excuse

| Setting                          | Required | Default | Description                                                |
| -------------------------------- | -------- | ------- | ---------------------------------------------------------- |
| `EXCUSE_MAXIMUM_AMOUNT`          | n        | 10      | Maximum number of excuses to request per command execution |
| `EXCUSE_PER_USER_RATE_LIMIT`     | n        | 20      | Number of excuses per time limit which user can request.   |
| `EXCUSE_PER_USER_RATE_TIME_SPAN` | n        | 20      | Per user rate limiting interval in seconds.                |

#### Calendar Facts

| Setting                                  | Required | Default | Description                                                       |
| ---------------------------------------- | -------- | ------- | ----------------------------------------------------------------- |
| `CALENDAR_FACTS_PER_USER_RATE_LIMIT`     | n        | 20      | Number of calendar facts per time limit which user can request.   |
| `CALENDAR_FACTS_PER_USER_RATE_TIME_SPAN` | n        | 20      | Per user rate limiting interval in seconds.                       |

#### Statistics

| Setting                          | Required | Default | Description                                            |
| -------------------------------- | -------- | ------- | ------------------------------------------------------ |
| `STATISTICS_SHOW_TOPMOST_N`      | n        | 5       | Number of users to show per command in `!stats` output |

#### Hints

| Setting                       | Required | Default | Description                                            |
| ----------------------------- | -------- | ------- | ------------------------------------------------------ |
| `HINTS_SNARKY_COMMENT_CHANCE` | n        | 100     | Reciprocal of chance of a snarky comment, rather than a helpful hint, being returned. Set to `0` to disable snarky comments. |

#### Emerald

| Setting                        | Required | Default | Description                                               |
| ------------------------------ | -------- | ------- | --------------------------------------------------------- |
| `EMERALD_BASE_URL`             | n        |         | Base URL of Emerald. Set to empty to disable integration. |
| `EMERALD_MAXIMUM_INPUT_LENGTH` | n        | 500     | Maximum size of input which to accept.                    |

#### Puma

| Variable            | Default value | Required | Description                                                                                                                                   |
| ------------------- | ------------- | -------- | --------------------------------------------------------------------------------------------------------------------------------------------- |
| `PUMA_LISTEN_IP`    | 0.0.0.0       | y        | IP which the application server will bind to. If you run this application outside of a Docker container, you will likely want to change this! |
| `PUMA_LISTEN_PORT`  | 8080          | y        | Port which the application server will bind to.                                                                                               |
| `PUMA_THREADS_MIN`  | 0             | y        | Initial number of threads to spawn per worker.                                                                                                |
| `PUMA_THREADS_MAX`  | 16            | y        | Maximum number of threads to spawn per worker.                                                                                                |
| `PUMA_WORKERS`      | 2             | y        | Number of worker processes to spawn.                                                                                                          |

#### Discord OAuth

| Setting                       | Required | Default                               | Description                                               |
| ----------------------------- | -------- | ------------------------------------- | --------------------------------------------------------- |
| `DISCORD_OAUTH_BASE_URL`      | n        | https://discordapp.com                | Base URL of Discord API                                   |
| `DISCORD_OAUTH_CLIENT_ID`     | n        |                                       | OAuth client ID                                           |
| `DISCORD_OAUTH_CLIENT_SECRET` | n        |                                       | OAuth client secret                                       |
| `DISCORD_OAUTH_CALLBACK_URL`  | n        | https://hivestalker.morrolan.ch/oauth | OAuth callback URL                                        |


CHANGELOG.md
# Change log

This document represents a high-level overview of changes made to this project.
It will not list every miniscule change, but will allow you to view - at a
glance - what to expact from upgrading to a new version.

## [unpublished]

### Added

### Changed

### Fixed

### Security

### Deprecated

### Removed


## [0.37.1] - 2020-11-01

### Changed

- Use `bundle exec` to start binaries within Docker container


## [0.37.0] - 2020-11-01

### Added

- Support for per-team and per-position skill
- `--debug` output for `!hive` command

### Changed

- Update bundler and dependencies.

### Removed

- Support for plaintext format for `!hive`


## [0.36.0] - 2019-11-10

### Added

- Allow automatically linking Discord-connected Steam account via `!link` 

### Changed

- Update ruby to `2.6.5`, update dependencies.


## [0.35.0] - 2019-11-03

### Added

- Sanitze input to `!hive`, reverting forced emoji shortcuts.


## [0.34.0] - 2019-05-09

### Changed

- Update dependencies, update base image to 2.6.3.
- Change `!tiplist guide` link to point to the wiki.

### Fixed

- `!latex` command was logged as `!version`.


## [0.33.0] - 2019-03-30

### Added

- `!latex` command to render simple LaTeX expressions.


## [0.32.0] - 2019-03-29

### Changed

- Add wiki link to `!tiplist movement`.
- Update dependencies.

### Fixed

- Typo in usage message of `!invite`.


## [0.31.0] - 2019-03-17

### Added

- Support for Gorge's cached player statistics.


## [0.30.0] - 2019-03-08

### Added

- Show guild ID in `!servers` output.


## [0.29.0] - 2019-03-08

### Added

- `!invite` command to create invite to one guild.


## [0.28.0] - 2019-03-04

### Changed

- Update dependencies. Most notably updates `hive-stalker` to `0.2.0`,
  supporting the new output format.


## [0.27.0] - 2019-02-26

### Added

- `!interblag` module https://xkcd.com/181/


## [0.26.0] - 2019-02-19

### Added

- `!leave` command which allows leaving a Discord guild.

### Changed

- Update dependencies
- Update base image to ruby 2.6.1


## [0.25.2] - 2019-02-04

### Changed

- Update dependencies


## [0.25.1] - 2018-11-03

### Changed

- Update dependencies.

### Fixed

- Allow placeholders containing numbers in hint texts.


## [0.25.0] - 2018-10-21

### Added

- Allow players to tell the bot about their Steam accounts.
- Use linked Steam account as default value for Hive queries.


## [0.24.0] - 2018-09-20

### Added

- Allow adding users to ignore list, which will make the bot ignore their commands.


## [0.23.1] - 2018-09-11

### Fixed

- `!stats` combining output of `!excuse` and `!calendarfact`.


## [0.23.0] - 2018-09-11

### Added

- `!calendarfact` command which provides one with a well-known piece of
  calendar trivia.


## [0.22.1] - 2018-09-09

### Changed

- Update dependencies and base image.

### Fixed

- `!hive` not supporting Steam ID64s outside of usual range.


## [0.22.0] - 2018-09-02

### Added

- `!tags` command which allows to list known hint tags, sorted alphabetically
  or by number of tagged hints.
- `!reloadhints` command which allows reloading hints from the source.

### Fixed

- Log usages of `!tipslist` command.


## [0.21.1] - 2018-08-31

### Fixed

- Placeholder replacement did not prefer longest matches, leading to partial
  replacements of placeholders if one placeholder was a substring of another.


## [0.21.0] - 2018-08-23

### Added

- `!tiplist` command which lists various sources of hints and tips.
- `!tip` command to retrieve hints and tips about NS2.
- Command to automatically import NS2 tips from Google doc.


## [0.20.0] - 2018-07-16

### Added

- Persistence layer using a Postgres database.
- Tracking of command usage per-user.
- `!stats` command to show top x users per command.


## [0.19.2] - 2018-06-26

### Fixed

- Gorge-based accuracy data was shown as fraction rather than percentage.


## [0.19.1] - 2018-06-25

### Fixed

- Exception if Gorge data incomplete, eg no accuracy available.


## [0.19.0] - 2018-05-19

### Changed

- Show no-onos marine accuracy.
- Update to ruby 2.5.1, update dependencies, update Tini.


## [0.18.0] - 2018-04-04

### Added

- Gorge integration, allowing to pull additional per-player statistics from
  Gorge.


## [0.17.0] - 2018-03-20

### Added

- Allow creating invites for servers the bot is in.

### Fixed

- Prevent timeout when listing servers due to counting members in server.


## [0.16.0] - 2018-03-15

### Added

- `!servers` command to list servers the bot is in, available to admin users.
- Server count exposed as Prometheus metric.


## [0.15.2] - 2018-03-09

### Changed

- Disable buffering of stdout.


## [0.15.1] - 2017-12-20

### Fixed

- Adjust to breaking changes in `steam-id2` library.


## [0.15.0] - 2017-12-19

### Added

- Bot will respond with an excuse if an internal exception is encountered while
  processing commands.

### Changed

- Update dependencies and base image.


## [0.14.0] - 2017-12-07

### Changed

- Improve plaintext and embed output formats of Hive queries.
  - Use markdown to improve readability of plaintext format.
  - Add Observatory link to plaintext format.
  - Decrease size of embed format.


## [0.13.0] - 2017-12-03

### Added

- Optional support for Discord-style "embeds" in Hive module.


## [0.12.0] - 2017-11-07

### Added

- Metrics tracking executed commands, as well as Hive API call timings.

### Changed

- Move list of excuses to separate file.


## [0.11.1] - 2017-11-01

### Fixed

- Prevent `!excuse` calls which exceeded the per-message limit from using up
  the rate limit.


## [0.11.0] - 2017-11-01

### Added

- Re-added per-message limit for `!excuse` command.
  Allows keeping message size in check, as well as prevent circumventing the
  rate limit upon the user first issuing the command.


## [0.10.0] - 2017-11-01

### Added

- Rate-limiting for `!excuse` command. Limits the number of excuses per time
  interval a user may request.

### Changed

- Change changelog link to Github.

### Removed


config/lerk.rb
# coding: utf-8

$stdout.sync = true

module Lerk
  module Config
    VALID_LOG_LEVELS = [:debug, :verbose, :normal, :quiet, :silent]

    DISCORD_CLIENT_ID   = ENV.fetch('DISCORD_CLIENT_ID')
    DISCORD_TOKEN       = ENV.fetch('DISCORD_TOKEN')

    LOG_LEVEL           = ENV.fetch('LOG_LEVEL', 'normal').to_sym
    unless VALID_LOG_LEVELS.include? LOG_LEVEL
      raise ArgumentError, "LOG_LEVEL '#{ LOG_LEVEL }' invalid. Valid values are: #{ VALID_LOG_LEVELS.join(', ') }"
    end

    module Lerk
      COMMAND_PREFIX = ENV.fetch('LERK_COMMAND_PREFIX', '!')

      ADMIN_USERS = ENV.fetch('LERK_ADMIN_USERS', '').split(',').map do |id|
        raise ArgumentError, "ADMIN_USERS must contain numeric IDs only!" unless id.match(/^\d+$/)
        id.to_i
      end

      HINTS_ADMIN_USERS = ENV.fetch('LERK_HINTS_ADMIN_USERS', '').split(',').map do |id|
        raise ArgumentError, "HINTS_ADMIN_USERS must contain numeric IDs only!" unless id.match(/^\d+$/)
        id.to_i
      end
    end

    module Prometheus
      PORT = ENV.fetch('PROMETHEUS_PORT', 5000).to_i
      LISTEN_IP = ENV.fetch('PROMETHEUS_LISTEN_IP', '0.0.0.0')
      ENABLED = ENV.fetch('PROMETHEUS_ENABLED', 'true') == 'true'
    end

    module HiveInterface
      STEAM_WEB_API_KEY   = ENV.fetch('STEAM_WEB_API_KEY', '')

      # Defaults to 2/s
      GLOBAL_RATE_LIMIT     = ENV.fetch('HIVE_GLOBAL_RATE_LIMIT', 2).to_i
      GLOBAL_RATE_TIME_SPAN = ENV.fetch('HIVE_GLOBAL_RATE_TIME_SPAN', 1).to_i

      # Defaults to 1/s
      PER_USER_RATE_LIMIT     = ENV.fetch('HIVE_USER_RATE_LIMIT', 1).to_i
      PER_USER_RATE_TIME_SPAN = ENV.fetch('HIVE_USER_RATE_TIME_SPAN', 1).to_i

      # Defaults to 1/5 min
      HELP_MESSAGE_RATE_LIMIT     = ENV.fetch('HIVE_HELP_MESSAGE_RATE_LIMIT', 1).to_i
      HELP_MESSAGE_RATE_TIME_SPAN = ENV.fetch('HIVE_HELP_MESSAGE_RATE_TIME_SPAN', 300).to_i

      PER_USER_RATE_LIMIT_MESSAGE = <<EOF
Hello there!

You have reached the rate limit of Hive 2 queries.
Please wait a few seconds before you issue another query. :)
EOF
    end

    module Observatory
      BASE_URL = ENV.fetch('OBSERVATORY_BASE_URL', 'https://observatory.morrolan.ch')
    end

    module Excuse
      MAXIMUM_AMOUNT = ENV.fetch('EXCUSE_MAXIMUM_AMOUNT', 10).to_i
      PER_USER_RATE_LIMIT     = ENV.fetch('EXCUSE_PER_USER_RATE_LIMIT', 20).to_i
      PER_USER_RATE_TIME_SPAN = ENV.fetch('EXCUSE_PER_USER_RATE_TIME_SPAN', 60).to_i
    end

    module CalendarFacts
      PER_USER_RATE_LIMIT     = ENV.fetch('CALENDAR_FACTS_PER_USER_RATE_LIMIT', 5).to_i
      PER_USER_RATE_TIME_SPAN = ENV.fetch('CALENDAR_FACTS_PER_USER_RATE_TIME_SPAN', 60).to_i
    end

    module Gorge
      BASE_URL            = ENV['GORGE_BASE_URL']
      HTTP_BASIC_USER     = ENV['GORGE_HTTP_BASIC_USER']
      HTTP_BASIC_PASSWORD = ENV['GORGE_HTTP_BASIC_PASSWORD']
      CONNECT_TIMEOUT     = ENV.fetch('GORGE_CONNECT_TIMEOUT', 1).to_i
      TIMEOUT             = ENV.fetch('GORGE_TIMEOUT', 2).to_i
      STATISTICS_CLASS    = ENV.fetch('GORGE_STATISTICS_CLASS', 'n_30')
    end

    module Statistics
      SHOW_TOPMOST_N = ENV.fetch('STATISTICS_SHOW_TOPMOST_N', '5').to_i
    end

    module Hints
      SNARKY_COMMENT_CHANCE = ENV.fetch('HINTS_SNARKY_COMMENT_CHANCE', '100').to_i
    end

    module Database
      ADAPTER  = 'postgres'
      HOST     = ENV['DB_HOST']
      PORT     = ENV['DB_PORT']
      DATABASE = ENV.fetch('DB_DATABASE')
      USER     = ENV['DB_USER']
      PASS     = ENV['DB_PASS']
    end

    module Emerald
      BASE_URL             = ENV.fetch('EMERALD_BASE_URL', '')
      MAXIMUM_INPUT_LENGTH = ENV.fetch('EMERALD_MAXIMUM_INPUT_LENGTH', '500').to_i
    end

    module DiscordOAuth
      CLIENT_ID     = ENV['DISCORD_OAUTH_CLIENT_ID']
      CLIENT_SECRET = ENV['DISCORD_OAUTH_CLIENT_SECRET']
      BASE_URL      = ENV.fetch('DISCORD_OAUTH_BASE_URL', 'https://discordapp.com')
      CALLBACK_URL  = ENV.fetch('DISCORD_OAUTH_CALLBACK_URL', 'https://hivestalker.morrolan.ch/oauth')

      def self.client
        unless CLIENT_ID && CLIENT_SECRET
          raise ArgumentError, 'Must configure DISCORD_OAUTH_CLIENT_ID and DISCORD_OAUTH_CLIENT_SOCKET to use this functionality.'
        end

        OAuth2::Client.new(
          CLIENT_ID,
          CLIENT_SECRET,
          site: BASE_URL,
          authorize_url: '/api/oauth2/authorize',
          token_url: '/api/oauth2/token',
        )
      end
    end
  end
end
